libs/vkd3d/command.c

```diff
@@ -8306,6 +8306,135 @@ static void d3d12_command_list_clear_uav(struct d3d12_command_list *list,
     }
 }
 
+static void d3d12_command_list_clear_uav_with_copy(struct d3d12_command_list *list,
+        const struct d3d12_desc_split *d, struct d3d12_resource *resource,
+        const struct vkd3d_clear_uav_info *args, uint32_t clear_value,
+        UINT rect_count, const D3D12_RECT *rects)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &list->device->vk_procs;
+    unsigned int miplevel_idx, layer_count, i;
+    struct vkd3d_scratch_allocation scratch;
+    VkCopyBufferToImageInfo2KHR copy_info;
+    VkBufferImageCopy2KHR copy_region;
+    VkDeviceSize scratch_buffer_size;
+    D3D12_RECT curr_rect, full_rect;
+    VkMemoryBarrier barrier;
+
+    d3d12_command_list_track_resource_usage(list, resource, true);
+    d3d12_command_list_end_current_render_pass(list, false);
+
+    assert(args->has_view);
+    assert(d3d12_resource_is_texture(resource));
+
+    if (d3d12_resource_is_texture(resource))
+        miplevel_idx = args->u.view->info.texture.miplevel_idx;
+    else
+        miplevel_idx = 0;
+
+    layer_count = args->u.view->info.texture.vk_view_type == VK_IMAGE_VIEW_TYPE_3D
+            ? d3d12_resource_desc_get_depth(&resource->desc, miplevel_idx)
+            : args->u.view->info.texture.layer_count;
+
+    full_rect.left = 0;
+    full_rect.right = d3d12_resource_desc_get_width(&resource->desc, miplevel_idx);
+    full_rect.top = 0;
+    full_rect.bottom = d3d12_resource_desc_get_height(&resource->desc, miplevel_idx);
+
+    if (rect_count)
+    {
+        scratch_buffer_size = 0;
+
+        for (i = 0; i < rect_count; i++)
+        {
+            if (rects[i].right > rects[i].left && rects[i].bottom > rects[i].top)
+            {
+                unsigned int w = rects[i].right - rects[i].left;
+                unsigned int h = rects[i].bottom - rects[i].top;
+                scratch_buffer_size = max(scratch_buffer_size, layer_count * w * h * sizeof(uint32_t));
```

Comment:
Seems a bit wasteful to allocate for many layers if we can just broadcast the same layer buffer copy.

Reply:
Added code to iterate over subresource layers now, still need this for 3D though.